File: 2020/day7/src/main.js
const fs = require("fs");

const content = fs.readFileSync(__dirname + "/input.txt", "utf-8");

const canContainBag = (name, node) => {
  return node.children.some(({ child }) => {
    if (child.bagName === name) {
      return true;
    }
    return canContainBag(name, child);
  });
};

const numberOfBagsItContains = (node, amount = { total: 0 }) => {
  node.children.forEach((node) => {
    amount.total += node.amount;
    for (let i = 0; i < node.amount; i++) {
      numberOfBagsItContains(node.child, amount);
    }
  });
  return amount.total;
};

let nodes = [];

const getNode = (name) => {
  const node = nodes.find((n) => n.bagName === name);
  if (node) {
    return node;
  }
  const newNode = {
    bagName: name,
    children: [],
  };

  nodes.push(newNode);
  return newNode;
};

content.split("\n").forEach((line) => {
  const [_, bagName, contents] = line.match(/(.*) bags contain (.*)/);

  const node = getNode(bagName);

  contents
    .split(", ")
    .filter((content) => content !== "no other bags.")
    .map((content) => content.match(/(\d) (.*) bag/))
    .forEach((match) => {
      const [_, amount, bagNameContent] = match;
      const contentNode = getNode(bagNameContent);
      node.children.push({
        child: contentNode,
        amount: parseInt(amount, 10),
      });
    });
});

const bagsThatCanContainShinyGold = nodes.filter((node) => {
  return canContainBag("shiny gold", node);
});

console.log(
  "Bags that can contain shiny gold",
  bagsThatCanContainShinyGold.length
);

const shinyGoldBag = nodes.find((node) => node.bagName === "shiny gold");

console.log("Shiny gold is a big bag: ", numberOfBagsItContains(shinyGoldBag));

File: 2020/day7/src/main.rs
use regex::Regex;
use std::fs;

struct Node {
    bag_name: String,
    children: Vec<Node>,
}

fn get_node<'a>(name: &str, nodes: &'a Vec<Node>) -> &'a Node {
    for node in nodes {
        if node.bag_name == name {
            return node;
        }
    }
    let newNode = Node {
        bag_name: name.to_string(),
        children: Vec::new(),
    };
    nodes.push(newNode);
    return &newNode;
}

fn part1(content: String) {
    let mut nodes: Vec<Node> = Vec::new();

    let re = Regex::new("(.*) bags contain (.*)").unwrap();
    let bagRe = Regex::new("\\d (.*) bags").unwrap();
    for line in content.lines() {
        let captures = re.captures(line).unwrap();
        let bag = &captures[1];
        let contents: Vec<&str> = captures[2].split(",").collect();

        let node = get_node(bag, &nodes);

        for content in contents {
            if bagRe.is_match(content) {
                let content_bag_name = &bagRe.captures(content).unwrap()[1];
                println!("Content {}", content_bag_name)
            }
        }
        println!("{}", bag);
    }
}

fn main() {
    let content =
        fs::read_to_string("./src/input.txt").expect("Something went wrong reading the file");

    part1(content);
}
